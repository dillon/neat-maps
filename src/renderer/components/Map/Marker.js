import React from 'react';
import PropTypes from 'prop-types';

const styles = {
  containerDefault: {
    position: 'absolute',
    transform: 'translate(-50%, -50%)',
    textDecoration: 'none',
    color: 'black',
  },
  containerHover: {
    position: 'absolute',
    transform: 'translate(-50%, -50%)',
    zIndex: 900,
    textDecoration: 'none',
    color: 'black',
  },
  dotDefault: ({ backgroundColor }) => ({
    width: 15,
    height: 15,
    backgroundColor,
    border: '1px solid black',
    borderRadius: '50%'
  }),
  dotHover: {
    width: 15,
    height: 15,
    backgroundColor: 'transparent',
    cursor: 'pointer',
    border: '1px solid black',
    borderRadius: '50%',
  },
  label: {
    padding: 10,
    border: '1px solid black',
    position: 'absolute',
    transform: 'translate(-50%, calc(-50%-50px)',
    minWidth: 150,
    backgroundColor: 'white',
  },
  categoryText: {
    marginTop: 5,
    fontStyle: 'italic'
  }
}

const Marker = ({ backgroundColor, $hover = false, text: { fullAddress, category } }) => {
  const { dotDefault, dotHover, label, containerHover, containerDefault, categoryText } = styles;
  const url = `https://www.google.com/maps/place/${fullAddress.replace(' ', '+')}`
  return (
    <a href={url} target="_blank" rel="noopener noreferrer" style={$hover ? containerHover : containerDefault}>
      <div
        style={$hover ? dotHover : dotDefault({ backgroundColor })}
      />
      {$hover &&
        <div style={label}>
          <div>
            {fullAddress}
          </div>
          <div style={categoryText}>
            {category}
          </div>
        </div>
      }
    </a>
  )
}

Marker.defaultProps = {
  $hover: false
}

Marker.propTypes = {
  backgroundColor: PropTypes.string.isRequired,
  $hover: PropTypes.bool,
  text: PropTypes.shape({
    fullAddress: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired
  }).isRequired
}

export default Marker;