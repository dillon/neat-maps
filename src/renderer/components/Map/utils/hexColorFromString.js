/* eslint-disable no-bitwise */

function hashCode(str) { // java String#hashCode
  let hash = 0;
  for (let i = 0; i < str.length; i += 1) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  return hash;
}

function intToRGB(i) {
  const c = (i & 0x00FFFFFF)
    .toString(16)
    .toUpperCase();

  return "00000".substring(0, 6 - c.length) + c;
}

export default (string) => `#${intToRGB(hashCode(string))}`