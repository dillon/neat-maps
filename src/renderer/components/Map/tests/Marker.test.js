import React from 'react';
import { create } from 'react-test-renderer';
import Marker from '../Marker';
import hexColorFromString from '../utils/hexColorFromString';

const row = {
  lat: 123.45678,
  long: -12.34567,
  category: 'this is an example category',
  fullAddress: '138 west 12th street, Queens, NY 10018'
}

it('renders correctly', () => {
  const { fullAddress, category, lat, lng } = row;
  const tree = create(
    <Marker
      key={`${lat}-${lng}`}
      lat={lat}
      lng={lng}
      backgroundColor={hexColorFromString(row.category)}
      text={{ fullAddress, category }}

    />
  )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
