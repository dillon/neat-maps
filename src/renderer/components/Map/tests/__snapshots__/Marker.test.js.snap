// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<a
  href="https://www.google.com/maps/place/138+west 12th street, Queens, NY 10018"
  rel="noopener noreferrer"
  style={
    Object {
      "color": "black",
      "position": "absolute",
      "textDecoration": "none",
      "transform": "translate(-50%, -50%)",
    }
  }
  target="_blank"
>
  <div
    style={
      Object {
        "backgroundColor": "#F75E53",
        "border": "1px solid black",
        "borderRadius": "50%",
        "height": 15,
        "width": 15,
      }
    }
  />
</a>
`;
